### ISSUE TITLE
[Feat] Bilateral filter

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#702]
## 🚀 Feature

Would be great to have a differentiable version of Bilater Filtering `kornia.filter.bilateral_blur`

We can follow opencv signature:
https://docs.opencv.org/3.4/d4/d86/group__imgproc__filter.html#ga9d7064d478c95d60003cf839430737ed

Resource: http://jamesgregson.ca/bilateral-filtering-in-python.html

**Comment by tvercaut** [comment]
Such a feature would be great. Especially if it also works in 3D and is implemented in CUDA (or rather running efficiently on the GPU). See e.g. for a starting point in CUDA:
https://github.com/opencv/opencv_contrib/blob/master/modules/cudaimgproc/src/cuda/bilateral_filter.cu
http://developer.download.nvidia.com/compute/DevZone/C/html_x64/Image_Processing.html#bilateralFilter

**Comment by edgarriba** [comment]
@tvercaut not sure we are going to support custom cuda kernels for now. I would go first for a naive solution using plane pytorch

**Comment by tvercaut** [comment]
OK, I played a bit with it. For a simple 449 x 359 image and a spatial kernel diameter of 9, I ended up with these timings (the pytorch based version is buggy but I would think the timing is roughly OK):
```
OpenCV CPU
100 loops, best of 3: 8.71 ms per loop

OpenCV UMat (OpenCL)
1000 loops, best of 3: 289 µs per loop

"Naive" numpy (Gregson based)
1 loop, best of 3: 793 ms per loop

"Naive" pytorch (Gregson based) - debugging required
10 loops, best of 3: 79.9 ms per loop
```
The CPU-GPU IO is not counted.

**Comment by edgarriba** [comment]
@tvercaut great ! consider submitting a PR. We can assist you to complete

**Comment by tvercaut** [comment]
Sorry, I don't have the bandwidth to do any real software development these days. If helpful, my quick and dirty notebook is here:
https://colab.research.google.com/drive/1hUsuHsCD1_QOhx7yrTfSsWsylmaKRu8b?usp=sharing

**Comment by edgarriba** [comment]
@tvercaut this is already an amazing contribution :)) thanks a lot

**Comment by AnimeshMaheshwari22** [comment]
Hello! I would like to work in the Bilateral filter functionality.

**Comment by lferraz** [comment]
Hi @AnimeshMaheshwari22 

> Hello! I would like to work in the Bilateral filter functionality.

perfect!!!! Let's see how much work requires this Feature and then we can determine how to proceed. :)

Luis

**Comment by edgarriba** [comment]
Agree, let’s continue this discussion in slack

**Comment by AhmadM-DL** [comment]
Any updates on this?

**Comment by edgarriba** [comment]
@AnimeshMaheshwari22 is going to give a try on that, @AhmadM-DL feel free to join our slack in case you want to be involved too in that feature.

**Comment by avani17101** [comment]
Don't seem to find any `kornia.filter.bilateral_blur` variable. Did its name change in new version?

**Comment by edgarriba** [comment]
@avani17101 bilateral file was not finished since we couldn't fins a proper API vs "fast" implementation. This particular algorithm needs to almost iterate through the different pixels to compute weighting in a local neighborhood.  This is extremely slow with python -- we should find another good vectored way to implement decently efficient. /cc @ducha-aiki

**Comment by AnimeshMaheshwari22** [comment]
Hello. This is my apprehension as well. The functionality I found close to this was uniform spatial bilateral filtering that is relatively faster, but that is not the exact requirement.

**Comment by ducha-aiki** [comment]
@edgarriba how about this implementation? https://github.com/sunny2109/bilateral_filter_Pytorch/blob/main/bilateral_filter.py
It seems one can just use unfold2d, no?

**Comment by edgarriba** [comment]
might fit !! @lferraz suggested something similar with unfolding

**Comment by AnimeshMaheshwari22** [comment]
Since I had submitted a PR, I can continue that and give it a try!

**Comment by edgarriba** [comment]
> Since I had submitted a PR, I can continue that and give it a try!

would be great to follow this new implementation

**Comment by gau-nernst** [comment]
I was looking for bilateral filter and saw this issue and the unfinished PR. I took a try, following ideas from sunny2109 implementations.

```python
import torch
import torch.nn.functional as F
from kornia.filters.gaussian import get_gaussian_kernel2d
import numpy as np
import cv2
import requests
import io
from PIL import Image
import matplotlib.pyplot as plt
from skimage.restoration import denoise_bilateral


def bilateral_filter(imgs, kernel_size, sigma_intensity, sigma_space):
    pad = (kernel_size - 1) // 2
    padded = F.pad(imgs, (pad, pad, pad, pad), mode="reflect")
    unfolded = padded.unfold(2, kernel_size, 1).unfold(3, kernel_size, 1).flatten(-2)  # (B, C, H, W, K x K)
    intensity_kernel = torch.exp(-0.5 * (unfolded - imgs[..., None]).square() / sigma_intensity ** 2)
    space_kernel = get_gaussian_kernel2d(
        kernel_size,
        (sigma_space, sigma_space),
        device=imgs.device,
        dtype=imgs.dtype
    ).view(1, 1, 1, 1, -1)
    kernel = space_kernel * intensity_kernel
    out = (unfolded * kernel).sum(-1) / kernel.sum(-1)
    return out


img = io.BytesIO(requests.get("https://upload.wikimedia.org/wikipedia/en/7/7d/Lenna_%28test_image%29.png").content)
x_np = np.array(Image.open(img))
x_np = x_np.astype(np.float32) / 255
x = torch.from_numpy(x_np).permute(2, 0, 1).unsqueeze(0)

out1 = bilateral_filter(x, 11, 0.5, 2)
out2 = cv2.bilateralFilter(x_np, 11, 0.5, 2)
out3 = denoise_bilateral(x_np, 11, 0.5, 2, bins=100_000, mode="reflect", channel_axis=-1)

diff = np.abs(out1.squeeze(0).permute(1, 2, 0).numpy() - out2) / out2
print(diff.mean(), diff.max(), diff.min())  # 0.015516709 2.511976 0.0

diff = np.abs(out1.squeeze(0).permute(1, 2, 0).numpy() - out3) / out3
print(diff.mean(), diff.max(), diff.min())  # 0.047820166 2.9121413 0.0

plt.subplot(1, 3, 1)
plt.imshow(out1.squeeze(0).permute(1, 2, 0).mul(255).numpy().astype(np.uint8))
plt.title("pytorch")

plt.subplot(1, 3, 2)
plt.imshow((out2 * 255).astype(np.uint8))
plt.title("opencv")

plt.subplot(1, 3, 3)
plt.imshow((out3 * 255).astype(np.uint8))
plt.title("skimage")

plt.show()
```

However, comparing with OpenCV and skimage implementations, the results are significantly different. Even OpenCV and skimage are not the same. OpenCV is the sharpest, followed by my PyTorch implementation, and then skimage. skimage has very strong ghosting artifacts.

![Figure_1](https://user-images.githubusercontent.com/26946864/221741288-ff3a3fcb-fe89-43cb-9829-3acc5cc5eda2.png)

@ducha-aiki @edgarriba If you can find out any problems with my implementation, I'm happy to try and fix.

OpenCV code: https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/bilateral_filter.dispatch.cpp and https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/bilateral_filter.dispatch.cpp. It's not very readable so I can't really understand what would lead to different results. I noticed that their spatial Gaussian kernel is limited by the radius, not by a square window. (`if (sqrt(i*i + j*j) > radius) continue`). This would make the values in the 4 corners of the filter a bit different.

https://github.com/opencv/opencv/blob/b3c9842a98a263cef9daaded8dfa20a82f30e06b/modules/imgproc/src/bilateral_filter.dispatch.cpp#L206-L207

Quality-wise, reducing intensity sigma (preserve more edges) in my PyTorch implementation can make the result closer to OpenCV's.

I haven't looked in skimage code in-depth yet.

**Comment by edgarriba** [comment]
> It's not very readable so I can't really understand what would lead to different results. I noticed that their spatial Gaussian kernel is limited by the radius, not by a square window. (`if (sqrt(i*i + j*j) > radius) continue`).

wondering whether we can solve that by clipping or similar. Have you benchmarked times ? It's fine to included this implementation, but might be worth to explore prototyping a triton example too. See: https://github.com/openai/triton/blob/main/python/tutorials/01-vector-add.py

not sure though how we solve then the auto differentiation; or we just have different modes. /cc @ducha-aiki @tvercaut @t-vi

**Comment by ducha-aiki** [comment]
@gau-nernst that is very clever implementation, I love it!

**Comment by gau-nernst** [comment]
Input is 512x512x3. CPU is Intel W-2295, GPU is RTX3090. Time measured to process 1 image. For batch 32, time is divided by 32.

Implementation | kernel_size=3 | kernel_size=5 | kernel_size=7
--|--|--|--
This PyTorch (CPU - batch 1) | 11.0294ms | 53.6026ms | 99.6424ms
This PyTorch (CPU - batch 32) | 19.2862ms | 46.0436ms | 83.7909ms
This PyTorch (GPU - batch 1) | 0.9488ms | 1.9329ms | 3.4727ms
This PyTorch (GPU - batch 32) | 0.6444ms | 1.6317ms | OOM
OpenCV | 0.2250ms | 0.4486ms | 1.4840ms
Scikit-image | 42.2278ms | 112.9555ms | 214.5870ms

It's not surprising that this is much slower than OpenCV. Doing calculations on unfolded tensors is expensive. In terms of speed, I can try optimize some operations e.g. use `torch.bmm()` at some places. Scikit-image being slower than PyTorch is quite surprising, considering that they use custom Cython code. Or perhaps I'm doing something wrong with scikit-image.

@edgarriba regarding the `if (sqrt(i*i + j*j) > radius) continue`, I think it's possible to create a boolean mask to mask out the Gaussian kernel before unfold it. I will try and let you know.

**Comment by edgarriba** [comment]
I would benchmark also against torchdynamo, we are seeing some speed improvements compared to opencv. /cc @johnnv1 https://github.com/kornia/kornia-benchmark/tree/master/dynamo

**Comment by ducha-aiki** [comment]
@edgarriba @gau-nernst I believe that better strategy would be to merge correct, but slower implementation first and do the speed-ups in the consequent PRs

**Comment by gau-nernst** [comment]
I just added speed test with `torch.compile()`. It's unbelievable!

@ducha-aiki Yes I agree, correctness first before speed optimization.

**Comment by edgarriba** [comment]
> I just added speed test with `torch.compile()`. It's unbelievable!

tell me more :)

**Comment by gau-nernst** [comment]
I think I have figured out the difference. For RGB images, OpenCV uses sum of absolute differences across three channels to calculate the intensity difference $|r_1-r_2| + |g_1-g_2| + |b_1-b_2|$

https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/bilateral_filter.simd.hpp#L880

The code above should be modified like this (take absolute value of the difference, then sum over channel dimension)

```python
    intensity_diff = (unfolded - imgs.unsqueeze(-1)).abs().sum(1, keepdim=True)  # (B, 1, H, W, K x K)
    intensity_kernel = torch.exp(-0.5 / sigma_intensity ** 2 * intensity_diff.square())
```

Numerically, the absolute difference is 0.12% on average, and 12.8% maximum. Visually, I think they look identical. Using the same settings as above (kernel size = 11, sigma intensity / color = 0.5, sigma space = 2)

![Figure_1](https://user-images.githubusercontent.com/26946864/221893553-5b6bdbb9-3c44-41e4-af99-592e6131751e.png)

The radial mask `if (sqrt(i*i + j*j) > radius) continue` can be implemented like the following

```python
    mask = torch.arange(-pad, pad+1, device=imgs.device).square()
    mask = (mask.view(-1, 1) + mask.view(1, -1)) <= kernel_size * kernel_size
    space_kernel = space_kernel * mask
```

I don't observe any difference when including the radial mask.

Also, regarding bilateral filter for RGB images, this paper (https://ieeexplore.ieee.org/document/710815) suggests converting to Lab color space and applying Euclidean distance (square root of sum of squared differences) $\sqrt{(L_1-L_2)^2 + (a_1-a_2)^2 + (b_1-b_2)^2}$. Matlab also suggests doing the same (https://www.mathworks.com/help/images/ref/imbilatfilt.html).

**Comment by tvercaut** [comment]
These may also be of interest:
https://github.com/faebstn96/trainable-bilateral-filter-source
https://github.com/faebstn96/trainable-joint-bilateral-filter-source
https://docs.monai.io/en/stable/networks.html#bilateralfilter

**Comment by gau-nernst** [comment]
> These may also be of interest: https://github.com/faebstn96/trainable-bilateral-filter-source https://github.com/faebstn96/trainable-joint-bilateral-filter-source https://docs.monai.io/en/stable/networks.html#bilateralfilter

I saw those custom C++/CUDA kernel code. I'm not familiar with writing such code so I probably won't be able to write it (unless we just directly copy it?) Compiling and packaging will be more difficult, and harder to maintain. For example, I tried compiling Monai (they have no pre-compiled kernels on PyPI yet) for my macbook. There were no errors in the compilation/installation phase but I couldn't call the bilateral filter function (I got some missing symbol errors).

Nevertheless, looking at these implementations, it seems like they use L2 distance for color difference (like Matlab). 

https://github.com/faebstn96/trainable-bilateral-filter-source/blob/main/csrc/bf_layer_gpu_forward.cu#L114

https://github.com/Project-MONAI/MONAI/blob/dev/monai/csrc/filtering/bilateral/bilateralfilter_cuda.cu#L111

**Comment by edgarriba** [comment]
I don’t think we have enough infrastructure/bandwidth to support custom c++/cuda . I’d explore triton instead

**Comment by gau-nernst** [comment]
I re-ran the speed benchmark for future reference.

Input is 16 x 3 x 256 x 256. CPU is Intel W-2295, GPU is RTX3090

device|kernel size|time|time (w/ `torch.compile()`)
---|---|---|---
cpu|3|0.233064|0.118785
cpu|7|0.716713|0.265270
cpu|11|1.339221|0.474534
cuda|3|0.002658|0.000469
cuda|7|0.011219|0.001648
cuda|11|0.025569|0.003081

```python
import time
import torch
import pandas as pd
from kornia.filters import bilateral_blur

torch.backends.cudnn.benchmark = True
torch.set_float32_matmul_precision('high')

N = 50
records = []

for device in ("cpu", "cuda"):
    x = torch.randn(16, 3, 256, 256, device=device)
    for compile in (False, True):
        fn = torch.compile(bilateral_blur) if compile else bilateral_blur
        for k in (3, 7, 11):
            print(f"{device = }, {compile = }, {k = }")

            # warmup
            for _ in range(2):
                out = fn(x, k, 0.1, (1, 1))

            time0 = time.perf_counter()
            for _ in range(N):
                if device == "cuda":
                    torch.cuda.synchronize()
                out = fn(x, k, 0.1, (1, 1))
                if device == "cuda":
                    torch.cuda.synchronize()
            _time = (time.perf_counter() - time0) / N

            records.append(dict(device=device, compile=compile, k=k, time=_time))

df = pd.DataFrame(records).to_csv("records.csv", index=False)
```
